### PR TITLE
Feature/gridedit 621 attach grid on boundary

### PR DESCRIPTION
**Pull request by BillSenior**

### CODE REVIEW

File: libs/MeshKernel/include/MeshKernel/Mesh2D.hpp [hunk 1]
```diff
@@ -352,5 +369,74 @@ namespace meshkernel
...
+        UInt FindOppositeEdge(const UInt faceId, const UInt edgeId) const;
+
+        /// @brief Get the next face adjacent to the edge on the opposite side.
+        /// @param [in] faceId The starting face
+        /// @param [in] edgeId The starting edge
+        /// @return Id of neighbour face along the edge
+        UInt NextFace(const UInt faceId, const UInt edgeId) const;
+
+        /// @brief Find all elements that do no have a neighbour across any of edges.
+        void GetElementsOnDomainBoundary(std::vector<UInt>& elementsOnDomainBoundary, std::vector<UInt>& edgesOnDomainBoundary) const;
```
Comment (lucacarniato) [inline libs/MeshKernel/include/MeshKernel/Mesh2D.hpp:388]: this is only for quads, either make this function generic (with the number of edges an element should have) or make it explicit on the name

File: libs/MeshKernel/src/Mesh2D.cpp [hunk 2]
```diff
@@ -1934,3 +1934,645 @@ std::vector<int> Mesh2D::NodeMaskFromPolygon(const Polygons& polygon, bool insid
...
+
+void Mesh2D::GatherHangingNodeIds(const std::vector<UInt>& edgesOnDomainBoundary,
+                                  std::vector<HangingNodeIndexArray>& irregularEdge,
+                                  std::vector<UInt>& irregularEdgeCount,
+                                  std::vector<UInt>& irregularStartNodes,
+                                  std::vector<UInt>& irregularEndNodes) const
+{
+    for (UInt i = 0; i < edgesOnDomainBoundary.size(); ++i)
+    {
+        UInt edgeI = edgesOnDomainBoundary[i];
```
Comment (lucacarniato) [inline libs/MeshKernel/src/Mesh2D.cpp:2112]: const auto edgeI

File: libs/MeshKernel/src/Mesh2D.cpp [hunk 3]
```diff
@@ -1934,3 +1934,645 @@ std::vector<int> Mesh2D::NodeMaskFromPolygon(const Polygons& polygon, bool insid
...
+            return m_edgesFaces[edgeId][0];
+        }
+    }
+
+    return constants::missing::uintValue;
+}
+
+void Mesh2D::AreEdgesAdjacent(const UInt edge1, const UInt edge2, bool& areAdjacent, UInt& startNode, UInt& endNode) const
+{
+    Point edge1Start = m_nodes[m_edges[edge1].first];
```
Comment (lucacarniato) [inline libs/MeshKernel/src/Mesh2D.cpp:2031]: const Point edge1Start = m_nodes[m_edges[edge1].first];
    const Point edge1End = m_nodes[m_edges[edge1].second];
    const Point edge2Start = m_nodes[m_edges[edge2].first];
    const Point edge2End = m_nodes[m_edges[edge2].second];

File: libs/MeshKernel/src/Mesh2D.cpp [hunk 4]
```diff
@@ -1934,3 +1934,645 @@ std::vector<int> Mesh2D::NodeMaskFromPolygon(const Polygons& polygon, bool insid
...
+                                  std::vector<UInt>& irregularStartNodes,
+                                  std::vector<UInt>& irregularEndNodes) const
+{
+    for (UInt i = 0; i < edgesOnDomainBoundary.size(); ++i)
+    {
+        UInt edgeI = edgesOnDomainBoundary[i];
+
+        for (UInt j = 0; j < edgesOnDomainBoundary.size(); ++j)
+        {
+            UInt edgeJ = edgesOnDomainBoundary[j];
```
Comment (lucacarniato) [inline libs/MeshKernel/src/Mesh2D.cpp:2116]: const UInt edgeJ = edgesOnDomainBoundary[j];

            if (i == j)
            {
                continue;
            }

            UInt startNode;
            UInt endNode;
            bool areAdjacent = false;

            AreEdgesAdjacent(edgeI, edgeJ, areAdjacent, startNode, endNode);

            if (!areAdjacent)
            {
                continue;
            }

            irregularEdgeCount[i] += 1;
            irregularEdge[i][irregularEdgeCount[i]] = j;

            if (startNode != constants::missing::uintValue)
            {
                irregularStartNodes[i] = startNode;
            }

            if (endNode != constants::missing::uintValue)
            {
                irregularEndNodes[i] = endNode;
            }

File: libs/MeshKernel/src/Mesh2D.cpp [hunk 5]
```diff
@@ -1934,3 +1934,645 @@ std::vector<int> Mesh2D::NodeMaskFromPolygon(const Polygons& polygon, bool insid
...
+            }
+        }
+    }
+}
+
+void Mesh2D::GatherNodesToMerge(const UInt startNode, const UInt endNode, const Edge& boundaryEdge, std::vector<std::pair<UInt, UInt>>& nodesToMerge, std::vector<int>& mergeIndicator) const
+{
+    if (startNode != constants::missing::uintValue && mergeIndicator[boundaryEdge.first] == 1)
+    {
+        nodesToMerge.push_back({boundaryEdge.first, startNode});
```
Comment (lucacarniato) [inline libs/MeshKernel/src/Mesh2D.cpp:2152]: nodesToMerge.emplace_back(boundaryEdge.first, startNode);

File: libs/MeshKernel/src/Mesh2D.cpp [hunk 6]
```diff
@@ -1934,3 +1934,645 @@ std::vector<int> Mesh2D::NodeMaskFromPolygon(const Polygons& polygon, bool insid
...
+    if (endNode != constants::missing::uintValue && mergeIndicator[boundaryEdge.second] == 1)
+    {
+        nodesToMerge.push_back({boundaryEdge.second, endNode});
+        mergeIndicator[boundaryEdge.second] = -1;
+    }
+}
+
+void Mesh2D::GatherHangingNodes(const UInt primaryStartNode, const UInt primaryEndNode, const Edge& irregularEdge, std::vector<UInt>& hangingNodesOnEdge, UInt& numberOfHangingNodes, std::vector<int>& mergeIndicator) const
+{
+    UInt secondaryStartNode = irregularEdge.first;
```
Comment (lucacarniato) [inline libs/MeshKernel/src/Mesh2D.cpp:2165]: const UInt secondaryStartNode = irregularEdge.first;
const UInt secondaryEndNode = irregularEdge.second;

File: libs/MeshKernel/src/Mesh2D.cpp [hunk 7]
```diff
@@ -1934,3 +1934,645 @@ std::vector<int> Mesh2D::NodeMaskFromPolygon(const Polygons& polygon, bool insid
...
+    std::vector<UInt> irregularEdgeCount(GetNumEdges(), missingValue);
+    std::vector<HangingNodeIndexArray> irregularEdge(GetNumEdges(), {missingValue, missingValue, missingValue, missingValue, missingValue});
+    std::vector<UInt> irregularStartNodes(GetNumEdges(), missingValue);
+    std::vector<UInt> irregularEndNodes(GetNumEdges(), missingValue);
+
+    GatherHangingNodeIds(edgesOnDomainBoundary, irregularEdge, irregularEdgeCount, irregularStartNodes, irregularEndNodes);
+
+    std::vector<bool> adjacentEdgeIndicator(GetNumEdges(), true);
+    std::vector<std::pair<UInt, UInt>> nodesToMerge;
+    nodesToMerge.reserve(4 * GetNumEdges());
```
Comment (lucacarniato) [inline libs/MeshKernel/src/Mesh2D.cpp:2207]: maybe use a named constant for 4. you can also put this variable in constants.hpp, namespace numerics

File: libs/MeshKernel/src/Mesh2D.cpp [hunk 8]
```diff
@@ -1934,3 +1934,645 @@ std::vector<int> Mesh2D::NodeMaskFromPolygon(const Polygons& polygon, bool insid
...
+
+    if (adjacentFaceId != constants::missing::uintValue)
+    {
+        UInt nextOppositeEdge = FindOppositeEdge(adjacentFaceId, edgeId);
+        ConnectNodes(newNodeIndex, m_edges[nextOppositeEdge].first);
+        ConnectNodes(newNodeIndex, m_edges[nextOppositeEdge].second);
+    }
+}
+
+void Mesh2D::FreeFourHangingNodes(const UInt faceId, const UInt edgeId, const HangingNodeIndexArray& hangingNodes, const UInt startNode, const UInt endNode)
```
Comment (lucacarniato) [inline libs/MeshKernel/src/Mesh2D.cpp:2397]: can you make all the dd boundary algorithm a separate class?

**Review by lucacarniato — CHANGES_REQUESTED** [review]

File: libs/MeshKernel/include/MeshKernel/Mesh2D.hpp (hunk 1 above)
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/Mesh2D.hpp:388]: Done

File: libs/MeshKernel/src/Mesh2D.cpp (hunk 2 above)
Comment (BillSenior) [reply libs/MeshKernel/src/Mesh2D.cpp:2112]: done

File: libs/MeshKernel/src/Mesh2D.cpp (hunk 3 above)
Comment (BillSenior) [reply libs/MeshKernel/src/Mesh2D.cpp:2031]: done

File: libs/MeshKernel/src/Mesh2D.cpp (hunk 4 above)
Comment (BillSenior) [reply libs/MeshKernel/src/Mesh2D.cpp:2116]: done

File: libs/MeshKernel/src/Mesh2D.cpp (hunk 5 above)
Comment (BillSenior) [reply libs/MeshKernel/src/Mesh2D.cpp:2152]: done

File: libs/MeshKernel/src/Mesh2D.cpp (hunk 6 above)
Comment (BillSenior) [reply libs/MeshKernel/src/Mesh2D.cpp:2165]: done

File: libs/MeshKernel/src/Mesh2D.cpp (hunk 7 above)
Comment (BillSenior) [reply libs/MeshKernel/src/Mesh2D.cpp:2207]: Used an already existing constant in the class.
I do not think this should be a globally visible constant, as it is only of use for this algorithm.

File: libs/MeshKernel/src/Mesh2D.cpp (hunk 8 above)
Comment (BillSenior) [reply libs/MeshKernel/src/Mesh2D.cpp:2397]: Done

**Review by ahmad-el-sayed — CHANGES_REQUESTED** [review]
@BillSenior  Do not be alarmed :) lots of const stuff.

File: libs/MeshKernel/include/MeshKernel/Point.hpp
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/Point.hpp:1]: Moved it to Utilities/NumericFunctions.hh

File: libs/MeshKernel/src/ConnectCurvilinearGrids.cpp [hunk 9]
```diff
@@ -0,0 +1,566 @@
...
+    endNode = constants::missing::uintValue;
+
+    if (edge1Start == edge1End || edge2Start == edge2End)
+    {
+        return;
+    }
+
+    const double edge1Length = ComputeDistance(edge1Start, edge1End, mesh.m_projection);
+    const double edge2Length = ComputeDistance(edge2Start, edge2End, mesh.m_projection);
+    const double minimumLength = 0.4 * std::min(edge1Length, edge2Length);
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/src/ConnectCurvilinearGrids.cpp:25]: I do not know, its from the f**tran code
Probably some close to, but less than a half

File: libs/MeshKernel/include/MeshKernel/ConnectCurvilinearGrids.hpp [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+{
+
+    /// @brief Connects curvilinear grids across element faces.
+    ///
+    /// Can be connected with upto 5 elements or 4 hanging nodes along an edge.
+    /// @note Currently only works for quadrilateral elements.
+    class ConnectCurvilinearGrids final
+    {
+    public:
+        /// @brief Connect curvilinear grids (connectcurvilinearquadsddtype.f90)
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/ConnectCurvilinearGrids.hpp:47]: I have found having the Fortran reference in a comment quite useful, I will keep it but not in the Doxygen comments

File: libs/MeshKernel/include/MeshKernel/Mesh2D.hpp [hunk 11]
```diff
@@ -293,14 +297,32 @@ namespace meshkernel
...
+        UInt FindOppositeEdge(const UInt faceId, const UInt edgeId) const;
+
+        /// @brief Get the next face adjacent to the edge on the opposite side.
+        /// @param [in] faceId The starting face
+        /// @param [in] edgeId The starting edge
+        /// @return Id of neighbour face along the edge
+        UInt NextFace(const UInt faceId, const UInt edgeId) const;
+
+        /// @brief Print the mesh in a form that can be loaded into matlab/octave.
+        void Print(std::ostream& out = std::cout) const;
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/Mesh2D.hpp:312]: done

File: libs/MeshKernel/include/MeshKernel/Point.hpp [hunk 12]
```diff
@@ -223,6 +223,11 @@ namespace meshkernel
     /// @returns \f$ p1.x = p2.x \wedge p1.y = p2.y)\f$
     bool IsEqual(const Point& p1, const Point& p2, const double epsilon);
 
+    /// @brief Compute the point at some position along the line connecting start- and end-point.
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/Point.hpp:226]: Done

File: libs/MeshKernel/src/ConnectCurvilinearGrids.cpp [hunk 13]
```diff
@@ -0,0 +1,566 @@
...
+
+    if (edge1Length <= edge2Length)
+    {
+        Point midPoint = 0.5 * (edge1Start + edge1End);
+
+        auto [distance, intersection, parameterisedDistance] = DistanceFromLine(midPoint, edge2Start, edge2End, mesh.m_projection);
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/src/ConnectCurvilinearGrids.cpp:31]: Done

File: libs/MeshKernel/src/ConnectCurvilinearGrids.cpp [hunk 14]
```diff
@@ -0,0 +1,566 @@
...
+        return;
+    }
+
+    const double edge1Length = ComputeDistance(edge1Start, edge1End, mesh.m_projection);
+    const double edge2Length = ComputeDistance(edge2Start, edge2End, mesh.m_projection);
+    const double minimumLength = 0.4 * std::min(edge1Length, edge2Length);
+
+    if (edge1Length <= edge2Length)
+    {
+        Point midPoint = 0.5 * (edge1Start + edge1End);
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/src/ConnectCurvilinearGrids.cpp:29]: Done

File: libs/MeshKernel/src/ConnectCurvilinearGrids.cpp [hunk 15]
```diff
@@ -0,0 +1,566 @@
...
+
+        auto [distance, intersection, parameterisedDistance] = DistanceFromLine(midPoint, edge2Start, edge2End, mesh.m_projection);
+        areAdjacent = distance != constants::missing::doubleValue && distance < minimumLength;
+    }
+    else
+    {
+        Point midPoint = 0.5 * (edge2Start + edge2End);
+
+        auto [distance, intersection, parameterisedDistance] = DistanceFromLine(midPoint, edge1Start, edge1End, mesh.m_projection);
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/src/ConnectCurvilinearGrids.cpp:38]: Done

File: libs/MeshKernel/src/ConnectCurvilinearGrids.cpp [hunk 16]
```diff
@@ -0,0 +1,566 @@
...
+        if (irregularEdgeCount[i] == missingValue)
+        {
+            continue;
+        }
+
+        UInt boundaryEdgeId = edgesOnDomainBoundary[i];
+        Edge boundaryEdge = mesh.m_edges[boundaryEdgeId];
+
+        UInt primaryStartNode = irregularStartNodes[i];
+        UInt primaryEndNode = irregularEndNodes[i];
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/src/ConnectCurvilinearGrids.cpp:208]: done

File: libs/MeshKernel/src/ConnectCurvilinearGrids.cpp [hunk 17]
```diff
@@ -0,0 +1,566 @@
...
+                {
+                    adjacentEdgeIndicator[boundaryEdgeId] = false;
+                    mesh.m_edges[boundaryEdgeId].first = missingValue;
+                    mesh.m_edges[boundaryEdgeId].second = missingValue;
+                }
+            }
+        }
+
+        UInt boundaryFaceId = elementsOnDomainBoundary[i];
+        Point boundaryNode = mesh.m_nodes[boundaryEdge.first];
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/src/ConnectCurvilinearGrids.cpp:238]: Done

File: libs/MeshKernel/src/ConnectCurvilinearGrids.cpp [hunk 18]
```diff
@@ -0,0 +1,566 @@
...
+    // Compute distances of nodes in list from the base point
+    for (UInt j = 0; j < numberOfNodes; ++j)
+    {
+        distance[j] = ComputeDistance(point, mesh.m_nodes[nodeIndices[j]], mesh.m_projection);
+    }
+
+    std::iota(distanceIndex.begin(), distanceIndex.begin() + numberOfNodes, 0);
+    std::sort(distanceIndex.begin(), distanceIndex.begin() + numberOfNodes, [&distance](UInt i, UInt j)
+              { return distance[i] < distance[j]; });
+    std::ranges::fill(nearestNeighbours, constants::missing::uintValue);
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/src/ConnectCurvilinearGrids.cpp:279]: Done

File: libs/MeshKernel/src/ConnectCurvilinearGrids.cpp [hunk 19]
```diff
@@ -0,0 +1,566 @@
...
+        }
+    }
+}
+
+void meshkernel::ConnectCurvilinearGrids::GatherHangingNodeIds(const Mesh2D& mesh,
+                                                               const std::vector<UInt>& edgesOnDomainBoundary,
+                                                               std::vector<BoundedIntegerArray>& irregularEdge,
+                                                               std::vector<UInt>& irregularEdgeCount,
+                                                               std::vector<UInt>& irregularStartNodes,
+                                                               std::vector<UInt>& irregularEndNodes) const
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/src/ConnectCurvilinearGrids.cpp:87]: The arrays can be different sizes

[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/src/ConnectCurvilinearGrids.cpp:87]: Phew! Done

File: libs/MeshKernel/include/MeshKernel/ConnectCurvilinearGrids.hpp [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+
+        /// @brief Get list of node id's ordered with distance from given point.
+        void GetOrderedDistanceFromPoint(const Mesh2D& mesh,
+                                         const std::vector<UInt>& nodeIndices,
+                                         const UInt numberOfNodes,
+                                         const Point& point,
+                                         BoundedIntegerArray& nearestNeighbours) const;
+
+        /// @brief Merge coincident nodes
+        void MergeNodes(Mesh2D& mesh, const std::vector<std::pair<UInt, UInt>>& nodesToMerge, std::vector<int>& mergeIndicator) const;
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/ConnectCurvilinearGrids.hpp:71]: Done

**Review by lucacarniato — APPROVED** [review]

**Review by ahmad-el-sayed — APPROVED** [review]

**Pull request merged**